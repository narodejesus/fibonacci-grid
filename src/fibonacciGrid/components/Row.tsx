import React, {useEffect, useState, memo} from 'react'

type Props = {
    onClick: () => void
    rowValue: number
    testId: string
}


enum RowState {
    empty = '',
    selected = 'selected',
    cleared = 'cleared',
    withValue = 'withValue'
}

const ANIMATION_TIMEOUT = 800
const useRowState = (rowValue: number): RowState => {
    const [rowState, setRowState] = useState<RowState>(RowState.empty)
    
    useEffect(() => {
        let timer
        if (rowValue !== 0) {
            setRowState(RowState.selected)
            timer = setTimeout(() => {
                setRowState(RowState.withValue)
            }, ANIMATION_TIMEOUT)
        } else {
            setRowState((prevState: RowState) => {
                if (prevState === '' && rowValue === 0) {
                    return
                } 
    
                return RowState.cleared
            })
    
            timer = setTimeout(() => {
                setRowState(RowState.empty)
            }, ANIMATION_TIMEOUT)
        }

        return () => {
            clearTimeout(timer)
        }
    }, [rowValue])

    return rowState
}

const Row = ({onClick, rowValue, testId}: Props): JSX.Element => {
    const rowState = useRowState(rowValue)

    return (
        <td 
            data-testid={testId}
            className={rowState} 
            onClick={onClick}>
                {rowValue || ''}
        </td>
    )
}

export default memo<Props>(Row)