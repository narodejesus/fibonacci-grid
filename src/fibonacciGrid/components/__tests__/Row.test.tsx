import React from 'react'
import Row from '../Row'
import {render} from '@testing-library/react'

jest.useFakeTimers()

describe('Row component', () => {
    it('renders row component', () => {
        const {baseElement} = render(
           <table>
                <tbody>
                    <tr>
                        <Row onClick={jest.fn()} rowValue={0} />
                    </tr>
                </tbody>
            </table>
        )

        expect(baseElement).toMatchSnapshot()
    })

    it('renders row component with rowValue', () => {
        const {baseElement} = render(
           <table>
                <tbody>
                    <tr>
                        <Row onClick={jest.fn()} rowValue={1} />
                    </tr>
                </tbody>
            </table>
        )

        expect(baseElement).toMatchSnapshot()
    })
})