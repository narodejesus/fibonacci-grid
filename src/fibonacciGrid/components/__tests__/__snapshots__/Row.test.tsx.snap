// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row component renders row component 1`] = `
<body>
  <div>
    <table>
      <tbody>
        <tr>
          <td />
        </tr>
      </tbody>
    </table>
  </div>
</body>
`;

exports[`Row component renders row component with rowValue 1`] = `
<body>
  <div>
    <table>
      <tbody>
        <tr>
          <td
            class="selected"
          >
            1
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</body>
`;
